content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+
+### Panicking
+
+A quick look at `twiggy` shows that the main contributors to the Wasm module size are functions related to string formatting and panicking (before LTO, there was also code for memory management in there). And that makes sense! The parameter `n` is unsanitized and used to index an array. Rust has no choice but to inject bounds checks. If a bounds check fails, Rust panics.
+
+One way to handle this is to do the bounds checking ourselves. Rust's compiler is really good at proving whether undefined behavior can happen or not.
+
+|||codediff|rust
+  fn nth_prime(n: usize) -> i32 {
++     if n < 0 || n >= PRIMES.len() { return -1; }
+      PRIMES[n]
+  }
+|||
+
+Arguably more idiomatic would be to lean into `Option<T>` APIs to control how the error case should be handled:
+
+|||codediff|rust
+  fn nth_prime(n: usize) -> i32 {
+-     PRIMES[n]
++     *PRIMES.get(n).unwrap_or(&-1)
+  }
+|||
+
+A third way would be to use some of the `unchecked` methods that Rust explicitly provides. These open the door to undefined behavior and as such are `unsafe`, but if you are okay carrying the burden to ensure safety, the gain in performance (or file size) can be significant!
+
+|||codediff|rust
+  fn nth_prime(n: usize) -> i32 {
+-     PRIMES[n]
++     unsafe { *PRIMES.get_unchecked(n) }
+  }
+|||
+
+We can try and stay on top of where we might cause a panic and try to handle those paths manually. However, once we start relying on third-party crates this is less and less likely to succeed, because we can't easily change how the library does its error handling internally.
+
+### LTO
+
+We’ll probably have to make our peace with the fact that once we can’t avoid having a code path for panics in our code base. While we can try and mitigate the impact of panics (and we will do that!), there is a rather powerful optimization that will often yield some significant code saving. This optimization pass is provided by LLVM and is called [LTO (Link-Time Optimization)][lto]. `rustc` compiles and then optimizes each crate that you use and then links it all together. However, there are certain optimizations that only become apparent during link time. A lot of functions have different branches depending on the input. During compile time, you can only see the invocations of the functions from within the same crate. At link time, you know _all_ possible invocations of any given functions, so it might be possible for the linker to prove that certain paths of a function will never be taken, allowing it that code branch or even an entire function.
+
+LTO is disabled by default, as it is quite a costly optimization that can slow down compile times significantly, especially in bigger crates. It can be enabled through one of `rustc`’s many [codegen options], which you control in the `profile` section of your `Cargo.toml`. Specifically, we need to add this line to our `Cargo.toml` to enable LTO in release builds:
```

Comment:
for more inclusive language
```suggestion
LTO is turned off by default, as it is quite a costly optimization that can slow down compile times significantly, especially in bigger crates. It can be enabled through one of `rustc`’s many [codegen options], which you control in the `profile` section of your `Cargo.toml`. Specifically, we need to add this line to our `Cargo.toml` to enable LTO in release builds:
```

Reply:
Oh I never thought about this word choice like that. Thank you!